cpp/src/quantiles/groupby.hpp

```diff
@@ -0,0 +1,146 @@
+
+  groupby(cudf::table const& key_table, bool include_nulls = false)
+  : _key_table(key_table)
+  , _num_keys(key_table.num_rows())
+  , _include_nulls(include_nulls)
+  {};
+
+  ~groupby() {
+    if (_key_sorted_order)
+      gdf_column_free(_key_sorted_order.get());
+    if (_unsorted_labels)
+      gdf_column_free(_unsorted_labels.get());
+  }
+
+  /**
+   * @brief Group and sort a column of values
+   * values within each group
+   * 
+   * Sorts and groups the @p val_col where the groups are dictated by key table
+   * and the elements are sorted ascending within the groups. Calculates and
+   * returns the number of valid values within each group.
+   * 
+   * @param val_col The value column to group and sort
+   * @return the sorted and grouped column and per-group valid count
+   */
+  std::pair<gdf_column, rmm::device_vector<gdf_size_type> >
+  sort_values(gdf_column const& val_col);
+
+  /**
+   * @brief Get a table of sorted unique keys
+   * 
+   * @return a new table in which each row is a unique row in the sorted key table.
+   */
+  cudf::table unique_keys();
+
+  /**
+   * @brief Get the number of groups in the key table
+   * 
+   */
+  gdf_size_type num_groups() const { return group_indices().size(); }
+
+  /**
+   * @brief Get the key sorted order.
+   * 
+   * @return the sort order indices for the key table.
+   *
+   * Gathering the key table by sort order indices will produce the sorted key table.
+   * 
+   * Computes and stores the key sorted order on first invocation, and returns the
+   * stored order on subsequent calls.
+   */
+  gdf_column const& key_sort_order();
+
+  /**
+   * @brief Get the group indices.
+   * 
+   * @return vector of indices of the starting point of each group in the sorted key table
+   * 
+   * Computes and stores the group indices on first invocation and returns
+   * the stored group indices on subsequent calls.
+   */
+  index_vector const& group_indices();
+
+  /**
+   * @brief Get the group labels
+   * 
+   * @return vector of group ID for each row in the sorted key column
+   * 
+   * Computes and stores labels on first invocation and returns stored labels on subsequent calls.
+   */
+  index_vector const& group_labels();
+
+  /**
+   * @brief Get the unsorted labels
+   * 
+   * @return column of group labels in the order of the unsorted key table
+   * 
+   * For each row in the key table, the unsorted label is the group it would belong to after sorting.
+   * contains the group it would belong to, after sorting
+   * 
+   * Computes and stores unsorted labels on first invocation and returns stored labels on subsequent calls.
+   */
+  gdf_column const& unsorted_labels();
+
+ private:
+
+  gdf_col_pointer     _key_sorted_order;
+  gdf_col_pointer     _unsorted_labels;
+  cudf::table const&  _key_table;
+
+  index_vec_pointer   _group_ids;
+  index_vec_pointer   _group_labels;
+
+  gdf_size_type       _num_keys;
```

Comment:
Actually, this is more than that. In case we want to ignore nulls, then `_num_keys` is less than `keys.num_rows()`. In that case, it stores the number of valid keys, and calculating that on demand is not a no-op.

Reply:
The keys can be null and still counted in `_num_valid_keys`. This is in the case where `include_nulls=true` and then we have to consider `NULL=NULL`

Reply:
It's being used in a lot of places and including the logic everywhere clutters things. Right now it mean the effective number of keys.